### PR TITLE
[esp/scene] initial support for Gibson semantic scene

### PR DESCRIPTION
**Pull request by msbaines**
This change include a Gibson semantic scene generator and parser.

## Motivation and Context

Required to support Gibson semantic mesh data.

## How Has This Been Tested

Wrote a simple test.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/esp/scene/GibsonSemanticScene.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+namespace esp {
+namespace scene {
+
+constexpr int kMaxIds = 10000; /* We shouldn't every need more than this. */
+
+bool SemanticScene::loadGibsonHouse(
+    const std::string& houseFilename,
+    SemanticScene& scene,
+    const quatf& worldRotation /* = quatf::Identity() */) {
+  if (!io::exists(houseFilename)) {
```
Comment (mosra) [inline src/esp/scene/GibsonSemanticScene.cpp]: `Cr::Utility::Directory::exists()`? IIRC the `io::` APIs are obsolete and should be gradually replaced with corrade's functionality.

File: src/esp/scene/GibsonSemanticScene.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "GibsonSemanticScene.h"
+#include "SemanticScene.h"
+
+#include <map>
+#include <string>
+
+#include <Corrade/Utility/String.h>
```
Comment (mosra) [inline src/esp/scene/GibsonSemanticScene.cpp]: Unused include?

**Comment by msbaines** [comment]
ping

All feedback addressed.

File: tools/npz2scn.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import numpy as np
+
+
+# Convert ndarrays to python lists so that we can serialize.
+# Tranform coordinates by rotating Y-axis to Z-axis
+def fixCoords(entry):
+    size = entry["size"].tolist()
+    size[1], size[2] = size[2], size[1]
+    entry["size"] = size
```
Comment (mathfac) [inline tools/npz2scn.py]: Looks like `entry["size"]` wasn't list, but now it is. Is that expected?

File: tools/npz2scn.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import argparse
+import json
+import sys
+
+import numpy as np
+
+
+# Convert ndarrays to python lists so that we can serialize.
+# Tranform coordinates by rotating Y-axis to Z-axis
+def fixCoords(entry):
```
Comment (mathfac) [inline tools/npz2scn.py]: If we know `entry` type we can add typing to the func, for example:
from typing import Any, Dict
def fixCoords(entry: Dict[str, Any]) -> None

File: tools/npz2scn.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import argparse
+import json
+import sys
+
+import numpy as np
+
+
+# Convert ndarrays to python lists so that we can serialize.
+# Tranform coordinates by rotating Y-axis to Z-axis
```
Comment (mathfac) [inline tools/npz2scn.py]:
```suggestion
# Transform coordinates by rotating Y-axis to Z-axis
```
Should it be:
`# Transform coordinates by rotating Y-axis to -Z-axis`

File: tools/npz2scn.py (hunk 4 above)
Comment (mathfac) [inline tools/npz2scn.py]: Based on naming convention:
```suggestion
def fix_coords(entry):
```

File: tools/npz2scn.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    size = entry["size"].tolist()
+    size[1], size[2] = size[2], size[1]
+    entry["size"] = size
+    loc = entry["location"].tolist()
+    loc[1], loc[2] = -loc[2], loc[1]
+    entry["location"] = loc
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Extract object IDs from npz.")
```
Comment (mathfac) [inline tools/npz2scn.py]:
```suggestion
    parser = argparse.ArgumentParser(description="Extracts object IDs from npz. Used for loading 3dscenegraph.stanford.edu semantic data.")
```

File: src/esp/scene/GibsonSemanticScene.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace esp {
+namespace scene {
+
+constexpr int kMaxIds = 10000; /* We shouldn't every need more than this. */
+
+bool SemanticScene::loadGibsonHouse(
+    const std::string& houseFilename,
+    SemanticScene& scene,
+    const quatf& worldRotation /* = quatf::Identity() */) {
```
[the remark this answers is not in the document]
Comment (bigbike) [reply src/esp/scene/GibsonSemanticScene.cpp]: The default arguments are only provided in the function declaration, not in the function definition.

File: tools/npz2scn.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    parser.add_argument("npz_path", metavar="foo.npz", help="path to .npz file to load")
+    parser.add_argument(
+        "scn_path", metavar="foo.scn", help="path to output file to write"
+    )
+    args = parser.parse_args()
+    data = np.load(args.npz_path, allow_pickle=True)["output"].item()
+    objects = data["object"]
+    for object in objects.values():
+        fixCoords(object)
```
Comment (apsdehal) [inline tools/npz2scn.py]: Use of keyword `object`, rename to `obj` maybe.

File: src/esp/scene/GibsonSemanticScene.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    object->index_ = id;
+
+    const std::string categoryName = jsonObject["class_"].GetString();
+    auto it = categories.find(categoryName);
+    if (it != categories.end()) {
+      object->category_ = scene.categories_[it->second];
+    } else {
+      int nextCategoryIndex = scene.categories_.size();
+      categories[categoryName] = nextCategoryIndex++;
+      auto category = std::make_shared<GibsonObjectCategory>(nextCategoryIndex,
```
Comment (bigbike) [inline src/esp/scene/GibsonSemanticScene.cpp]: nextCategoryIndex has been increased by 1.
It is NOT equal to categories[categoryName] now. Make sure if this is really what you want?

**Comment by msbaines** [comment]
I've now addressed all feedback.

**Pull request merged**